Document the NO_OP TariScript vulnerability (#2739)
tari-project/tari: docs/src/tari_script_no_op_vulnerability.md

# TariScript No-Op script vulnerability

This document describes a vulnerability in the current version of TariScript found by [@mikethetike](https://github.com/mikethetike).

The initial specification of how TariScript is incorporated into a Tari Mimblewimble (MW) style transaction is defined in the Tari [RFC-0201](https://rfc.tari.com/RFC-0201_TariScript.html). 

> This vulnerability applies to transactions where a Sender is spending one of their inputs that is bound by a NO_OP script like is suggested in [RFC-0201](https://rfc.tari.com/RFC-0201_TariScript.html#standard-mw-transaction)

Reviewer: ```suggestion
This vulnerability applies to transactions where a Sender is spending one or more their inputs where all the inputs are bound by a NO_OP script like is suggested in [RFC-0201](https://rfc.tari.com/RFC-0201_TariScript.html#standard-mw-transaction)
```

Reviewer: This is only a problem where all the inputs and all the outputs are changeable. 

if 
\\( \gamma = k_{sa1} + k_{sa2} - k_{Ob}U_b  \\)
Then Bob cant construct a k_{Ob} to satisfy (k_{sa1} + k_{sa2})

Author: Sure updated the wording

there are no operations in the script that use data that Bob is not aware of. This means that Bob could choose a new set of input data \\( \hat{\theta_a} \\) which contains a public script
key of his choosing, \\( \hat{K_{sa}} \\) to which he knows the private key \\( \hat{k_{sa}} \\). The execution of the NO_OP script for Alice's input will now leave \\( \hat{K_{sa}} \\) on 
the stack. Bob can choose a new nonce \\( \hat{R_{sa}} \\) and then produce a valid script signature (\\( \hat{s_{sa}}, \hat{R_{sa}} \\)) for the challenge 
\\( H( \hat{R_{sa}} || \alpha_a || \hat{\theta_a} || h) \\).

Bob can also now produce a new transaction script offset while choosing a new UTXO script offset for his UTXO, \\( \hat{k_{Ob}} \\). This means he can modify the script attached 
to his UTXO, \\( \hat{\sigma_{b}} \\), calculate a new modified commitment and range proof \\( \hat{RP_{b}} \\) for his new script and commit to it in a new transaction script offset

\\( \hat{\gamma} =  \hat{k_{sa}} - \hat{k_{Ob}} \hat{U_b} \\).

> This attack is only possible if Alice did not have a change output in the transaction because in that case Bob would not know the private script offset for the change output, \\( k_{oc} \\), to produce 

Reviewer: ```suggestion
This attack is only possible if Alice did not have a change output that he does not know the blinding factor for in the transaction because in that case Bob would not know the private script offset for the change output, \\( k_{oc} \\), to produce 
```

This attack is only open are ALL inputs are NO_OP and Bob knows the blingfactor for ALL outputs.

Author: lol, I am not sure how to measure the `bling` of an output. 

But I will update the wording to include the idea that all the inputs need to be NO_OP and the change output. I will briefly mention that its all outputs but I prefer to keep the cleaner paradigm of 1 change output and 1 receiver output are the outputs that can be in a Tx.
